File: contracts/light_client.py
from web3 import Web3, HTTPProvider
from solcx import compile_source, install_solc, set_solc_version, get_solc_version

from web3 import Web3

import json

# Function to inicialize the connection to the node
def connect(provider_url):
    # Initialize endpoint URL
    node_url = provider_url #'http://localhost:7545'

    # Create the node connection
    provider = Web3(Web3.HTTPProvider(node_url))

    # Verify if the connection is successful
    if provider.is_connected():
        print("-" * 50)
        print("Connection Successful")
        print("-" * 50)
        return provider
        
    else:
        print("Connection Failed")
        return None

# Function to get the contract
def get_contract(provider, contract_address, abi_json_path):

    with open(abi_json_path) as f:
        data = json.load(f)
        abi = data['abi']

    # Create smart contract instance
    contract = provider.eth.contract(address=contract_address, abi=abi)
    return contract

def send_transaction_to_contract(provider, contract, caller, private_key, pk):
    # initialize the chain id, we need it to build the transaction for replay protection
    chain_id = provider.eth.chain_id
    # Initialize address nonce
    nonce = provider.eth.get_transaction_count(caller)
    # Call your function
    call_function = contract.functions.addElement(pk).build_transaction({"chainId": chain_id, "from": caller, "nonce": nonce})

    # Sign transaction
    signed_tx = provider.eth.account.sign_transaction(call_function, private_key=private_key)

    # Send transaction
    send_tx = provider.eth.send_raw_transaction(signed_tx.rawTransaction)

    return send_tx

def fetch_list(contract):
    # Read the list - call means we are reading from the blockchain
    list = contract.functions.getList().call()
    return list

def query_pk_to_index(contract, pk):
    # Read the index - call means we are reading from the blockchain
    index = contract.functions.pkToIndex(pk).call()
    return index

File: contracts/light_client.js
const {Web3} = require('web3');

const ABI = require('./truffle/build/contracts/Boquila.json'); // SET HERE ABI of the contract
     
// SET UP THE WEB3 PROVIDER
let web3 = new Web3('http://localhost:7545');

let contract_address = '0x961B50CF253B974BEA3110BbeE4A81414D8c3ae5'; // PASTE THE CONTRACT ADDRESS HERE

let contract = new web3.eth.Contract(ABI.abi, contract_address);

console.log(contract.methods);

async function getDefaultAccaunt() {    
        // Get the accounts to use one as default to sign transactions 
        const accounts = await web3.eth.getAccounts();
        return accounts[1];
}

async function addElement(element, defaultAccount) {
    const result = await contract.methods.addElement(element).send({from: defaultAccount, gas: 3000000});
    console.log("Transaction Hash: " + result.transactionHash)
}

async function getList() {
    const list = await contract.methods.getList().call();
    return list;
}

async function getIndexOfElement(element) {
    const index = await contract.methods.getIndexOfElement(element).call();
    return index;
}

async function interact() {
    const defaultAccount = await getDefaultAccaunt();
    await addElement("Element1", defaultAccount);
    await addElement("Element2", defaultAccount);
    await addElement("Element3", defaultAccount);
    await addElement("Element4", defaultAccount);
    await addElement("Element5", defaultAccount);
    const list = await getList();
    console.log("List: " + list);
    const index = await getIndexOfElement("Element3");
    console.log("Index of Element3: " + index);
}

interact(); 